// @vkontakte/vk-mini-apps-router/dist/hoc/withRouter.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useActiveVkuiLocation } from '../hooks/useActiveVkuiLocation';
import { getDisplayName } from '../utils/utils';
import { useParams, usePopout, useRouteNavigator } from '../hooks/hooks';
import { useFirstPageCheck } from '../hooks/useFirstPageCheck';
import { useSearchParams } from '../hooks/useSearchParams';
/**
 * HOC для добавления свойств
 *
 * location: {@link ActiveVkuiLocationObject}
 *
 * routeNavigator: {@link RouteNavigator}
 *
 * popout: {@link JSX.Element}
 *
 * params: {@link Params}
 *
 * isFirstPage: boolean
 *
 * searchParams: {@link URLSearchParams}
 *
 * setSearchParams: {@link SetURLSearchParams}
 *
 * в переданный компонент
 *
 * ```typescript
 * export default withRouter(App);
 * ```
 * @param Component
 */
export function withRouter(Component) {
    function WithRouter(props) {
        const [searchParams, setSearchParams] = useSearchParams();
        const routerProps = {
            routeNavigator: useRouteNavigator(),
            location: useActiveVkuiLocation(),
            popout: usePopout(),
            params: useParams(),
            isFirstPage: useFirstPageCheck(),
            searchParams,
            setSearchParams,
        };
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        const propsWithRouter = {
            ...props,
            ...routerProps,
        };
        return _jsx(Component, { ...propsWithRouter });
    }
    WithRouter.displayName = `WithRouter(${getDisplayName(Component)})`;
    return WithRouter;
}
