// @vkontakte/vk-mini-apps-router/dist/page-types/TabPage.js
import { uniqueKey } from './common';
class BaseTabPage {
    constructor(id, path, modals) {
        this.id = id;
        this.path = path;
        this.modals = modals;
        modals.forEach((modal) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-expect-error
            this[uniqueKey(this, modal.id)] = modal;
        });
    }
    getRoutes() {
        return this.modals
            .map((modal) => modal.getRoutes())
            .flat()
            .map((route) => ({ ...route, tab: this.id }))
            .concat({
            path: this.path,
            tab: this.id,
        });
    }
}
export class TabPage extends BaseTabPage {
    constructor(id, path, modals = []) {
        super(id, path, modals);
        this.path = path;
        this.hasParams = false;
    }
}
export class TabPageWithParams extends BaseTabPage {
    constructor(id, path, paramKeys, modals = []) {
        super(id, path, modals);
        this.id = id;
        this.path = path;
        this.paramKeys = paramKeys;
        this.hasParams = true;
    }
}
export function createTab(id, path, modals, paramKeys) {
    if (paramKeys) {
        return new TabPageWithParams(id, path, paramKeys, modals ?? []);
    }
    return new TabPage(id, path, modals ?? []);
}
