// @vkontakte/vkjs/lib/IOSDetections.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    IOS_NO_KEYBOARD_ALLOWED_OFFSET: function() {
        return IOS_NO_KEYBOARD_ALLOWED_OFFSET;
    },
    IPHONE_KEYBOARD_REJECT_OFFSET: function() {
        return IPHONE_KEYBOARD_REJECT_OFFSET;
    },
    IPHONE_SAFARI_BOTTOM_BAR: function() {
        return IPHONE_SAFARI_BOTTOM_BAR;
    },
    IPHONE_X_SAFARI_BOTTOM_BAR: function() {
        return IPHONE_X_SAFARI_BOTTOM_BAR;
    },
    checkIPadOS: function() {
        return checkIPadOS;
    },
    detectIOS: function() {
        return detectIOS;
    },
    iosMajor: function() {
        return iosMajor;
    },
    iosMinor: function() {
        return iosMinor;
    },
    isIOS: function() {
        return isIOS;
    },
    isIOSChrome: function() {
        return isIOSChrome;
    },
    isIPad: function() {
        return isIPad;
    },
    isIPadOS: function() {
        return isIPadOS;
    },
    isIPhone: function() {
        return isIPhone;
    },
    isIPhoneX: function() {
        return isIPhoneX;
    },
    isLandscapePhone: function() {
        return isLandscapePhone;
    },
    isScrollBasedViewport: function() {
        return isScrollBasedViewport;
    },
    isWKWebView: function() {
        return isWKWebView;
    }
});
var _dom = require("./dom");
var IPHONE_SAFARI_BOTTOM_BAR = 45;
var IPHONE_X_SAFARI_BOTTOM_BAR = 85;
var IPHONE_KEYBOARD_REJECT_OFFSET = 180;
// 44 iPhone, 55 iPad, iPad Pro 69
var IOS_NO_KEYBOARD_ALLOWED_OFFSET = 70;
function detectIOS(ua) {
    if (!ua) {
        ua = _dom.canUseDOM ? navigator.userAgent : "";
    }
    ua = ua.toLowerCase();
    var isIPadOS = checkIPadOS(ua);
    var isIPad = isIPadOS || ua.includes("ipad");
    var isIPhone = !isIPad && ua.search(/iphone|ipod/) !== -1;
    var isIOS = isIPhone || isIPad;
    var iosVersion = isIOS && ua.match(/os ([\d_]+) like mac os x/i);
    var iosMajor = 0;
    var iosMinor = 0;
    if (isIPadOS) {
        iosMajor = 13;
        iosMinor = 0;
    } else if (iosVersion) {
        iosVersion = iosVersion[1].split("_");
        iosMajor = +iosVersion[0];
        iosMinor = +iosVersion[1];
    }
    iosVersion = null;
    var isScrollBasedViewport = iosMajor < 13 && !(iosMajor === 11 && iosMinor < 3);
    var isWKWebView = isIOS && checkWKWebView(ua);
    var isIPhoneX = false;
    if (_dom.canUseDOM) {
        isIPhoneX = isIOS && screen.width === 375 && screen.height === 812 && window.devicePixelRatio === 3;
    }
    var isIOSChrome = ua.search(/crios/i) !== -1;
    return {
        isIPad: isIPad,
        isIPhone: isIPhone,
        isIOS: isIOS,
        isIPadOS: isIPadOS,
        iosMajor: iosMajor,
        iosMinor: iosMinor,
        isWKWebView: isWKWebView,
        isScrollBasedViewport: isScrollBasedViewport,
        isIPhoneX: isIPhoneX,
        isIOSChrome: isIOSChrome
    };
}
var detect = /*#__PURE__*/ detectIOS();
var isIPad = /*#__PURE__*/ function() {
    return detect.isIPad;
}();
var isIPhone = /*#__PURE__*/ function() {
    return detect.isIPhone;
}();
var isIOS = /*#__PURE__*/ function() {
    return detect.isIOS;
}();
var isIPadOS = /*#__PURE__*/ function() {
    return detect.isIPadOS;
}();
var iosMajor = /*#__PURE__*/ function() {
    return detect.iosMajor;
}();
var iosMinor = /*#__PURE__*/ function() {
    return detect.iosMinor;
}();
var isWKWebView = /*#__PURE__*/ function() {
    return detect.isWKWebView;
}();
var isScrollBasedViewport = /*#__PURE__*/ function() {
    return detect.isScrollBasedViewport;
}();
var isIPhoneX = /*#__PURE__*/ function() {
    return detect.isIPhoneX;
}();
var isIOSChrome = /*#__PURE__*/ function() {
    return detect.isIOSChrome;
}();
function isLandscapePhone() {
    return Math.abs(window.orientation) === 90 && !isIPad;
}
// Reference:
// https://stackoverflow.com/questions/28795476/detect-if-page-is-loaded-inside-wkwebview-in-javascript/30495399#30495399
function checkWKWebView(ua) {
    if (!_dom.canUseDOM) {
        return false;
    }
    var webkit = window.webkit;
    if (webkit && webkit.messageHandlers) {
        return true;
    }
    var lte9 = /constructor/i.test(String(window.HTMLElement));
    var idb = !!window.indexedDB;
    if (ua.includes("safari") && ua.includes("version") && !navigator.standalone) {} else // Safari (WKWebView/Nitro since 6+)
    if (!idb && lte9 || !(window.statusbar && window.statusbar.visible)) {} else // UIWebView
    if (!lte9 || idb) {
        // WKWebView
        return true;
    }
    return false;
}
/**
 * В Safari на iPadOS поле User Agent содержит почти такую же информацию, что и в Safari на MacOS.
 * Из-за чего мы не можем ориентироваться на User Agent.
 *
 * Вместо этого мы пробуем определять есть ли событие 'ontouchend' в объекте document.
 *
 * см. https://developer.apple.com/forums/thread/119186?answerId=705140022#705140022
 */ function checkIPadOS(ua) {
    if (!_dom.canUseDOM) {
        return false;
    }
    var isNotIOS = !/ipPad|iPhone|iPod/i.test(ua);
    var isMacOS = /Mac OS/i.test(ua);
    return isNotIOS && isMacOS && "ontouchend" in document;
}
