// @vkontakte/vkjs/lib/type_checkers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    isArray: function() {
        return isArray;
    },
    isFormData: function() {
        return isFormData;
    },
    isFunction: function() {
        return isFunction;
    },
    isNumber: function() {
        return isNumber;
    },
    isObject: function() {
        return isObject;
    },
    isObjectLike: function() {
        return isObjectLike;
    },
    isString: function() {
        return isString;
    },
    isUndefined: function() {
        return isUndefined;
    }
});
function isObjectLike(object) {
    return typeof object === "object" && object !== null;
}
function isArray(object) {
    return Array.isArray(object);
}
function isObject(object) {
    return Object.prototype.toString.call(object) === "[object Object]";
}
function isUndefined(object) {
    return typeof object === "undefined";
}
function isFunction(object) {
    return typeof object === "function";
}
function isFormData(object) {
    return object && Object.prototype.toString.call(object) === "[object FormData]";
}
function isString(object) {
    return typeof object === "string";
}
function isNumber(object) {
    return typeof object === "number";
}
