// @vkontakte/vk-mini-apps-router/dist/index.js
export { RouterProvider } from './components/RouterProvider';
export { RouterLink } from './components/RouterLink';
export { createHashRouter } from './utils/createHashRouter';
export { createHashParamRouter } from './utils/createHashParamRouter';
export { createBrowserRouter } from './utils/createBrowserRouter';
export { useRouteNavigator, useParams, usePopout } from './hooks/hooks';
export { useGetPanelForView } from './hooks/useGetPanelForView';
export { useSearchParams } from './hooks/useSearchParams';
export { useMetaParams } from './hooks/useMetaParams';
export { useFirstPageCheck } from './hooks/useFirstPageCheck';
export { useActiveVkuiLocation } from './hooks/useActiveVkuiLocation';
export { useEnableSwipeBack } from './hooks/useEnableSwipeBack';
export { useHref } from './hooks/useHref';
export { useBlocker } from './hooks/useBlocker';
export { useLinkClickHandler } from './hooks/useLinkClickHandler';
export { withRouter } from './hoc/withRouter';
export { getInitialLocation } from './services/InitialLocation';
export { RoutesConfig, createView, createRoot, createPanel, createTab, createModal, } from './page-types';
